monkey/monkey_island/cc/services/agent_plugin_service/mongo_agent_plugin_repository.py

```diff
@@ -0,0 +1,226 @@
+            raise StorageError(
+                f"Plugin {plugin_name} of type {plugin_type} already has a binary for "
+                f"operating system {operating_system}"
+            )
+
+        try:
+            id = self._agent_plugins_binaries_collections[operating_system].put(
+                agent_plugin.source_archive
+            )
+            plugin_dict[BINARY_OS_MAPPING_KEY][operating_system.value] = id
+        except Exception:
+            raise StorageError(
+                f"Failed to store binary for plugin {plugin_name} of type {plugin_type} for "
+                f"operating system {operating_system}"
+            )
+
+        print(f"Writing plugin {plugin_dict}")
+        self._agent_plugins_collection.update_one(
+            {"plugin_manifest.name": plugin_name, "plugin_manifest.plugin_type": plugin_type.value},
+            {"$set": plugin_dict},
+            upsert=True,
+        )
+
+    def _encode_agent_plugin(self, agent_plugin: AgentPlugin) -> Dict[str, Any]:
+        dict = agent_plugin.dict(simplify=True, exclude={"source_archive"})
+
+        return self._encode_dict(dict)
+
+    def _encode_dict(self, d: dict):
+        new_dict = {}
+        for k, v in d.items():
+            if isinstance(k, Enum):
+                k = k.value
+            if isinstance(v, Enum):
+                v = v.value
+            if isinstance(v, dict):
+                v = self._encode_dict(v)
+
+            new_dict[k] = v
+
+        return new_dict
+
+    def remove_agent_plugin(
+        self,
+        agent_plugin_type: AgentPluginType,
+        agent_plugin_name: str,
+        operating_system: Optional[OperatingSystem] = None,
+    ):
+        try:
+            plugin_dict = self._get_agent_plugin(agent_plugin_type, agent_plugin_name)
+        except Exception:
+            print(f"Plugin {agent_plugin_name} of type {agent_plugin_type} not found")
+            return
+
+        try:
+            self._remove_agent_plugin(plugin_dict, operating_system)
+        except Exception as err:
+            raise RemovalError(
+                f"Error removing the agent plugin {agent_plugin_name} of type {agent_plugin_type} "
+                f"for operating system {operating_system}: {err}"
+            )
+
+    def _remove_agent_plugin(
+        self, plugin_dict: Dict[str, Any], operating_system: Optional[OperatingSystem]
+    ):
+        os_binaries = plugin_dict[BINARY_OS_MAPPING_KEY]
+
+        if operating_system is None:
+            temp_os_binaries = os_binaries.copy()
+            for os, id in temp_os_binaries.items():
+                self._agent_plugins_binaries_collections[OperatingSystem(os)].delete(id)
+                os_binaries.pop(os)
+        else:
+            id = os_binaries[operating_system.value]
+            self._agent_plugins_binaries_collections[OperatingSystem(operating_system)].delete(id)
+            os_binaries.pop(operating_system.value)
```

Comment:
Can we do something like:

```suggestion
        if operating_system is None:
            os_binaries_to_remove = os_binaries.copy()
        else:
            os_binaries_to_remove = [operating_system.value]
            
        for os, id in os_binaries_to_remove.items():
            self._agent_plugins_binaries_collections[OperatingSystem(os)].delete(id)
            os_binaries.pop(os)
```
^ WARNING: Not tested

Reply:
Yes, we should be able to do that